handbook/Best practices/TypeScript/Project-References.md

```diff
@@ -0,0 +1,166 @@
+# Project References
+
+## Table of contents
+
+1. [What are Project References?](#what-are-project-references)
+1. [Migrating a codebase](#migrating-a-codebase)
+1. [Tips](#tips)
+1. [Caveats](#caveats)
+
+## What are Project References?
+
+[Project References](https://www.typescriptlang.org/docs/handbook/project-references.html) provides built-in scalability for TypeScript. 
+
+### Code strucutre
+
+Project references provide developers tools to separate their code into smaller blocks. This enforces logical guidelines across the codebase which results in healthier code over time. 
+
+### Performance
+
+The logical guidelines enforced by Project References enable [incremental compiles](https://www.typescriptlang.org/docs/handbook/project-references.html#build-mode-for-typescript). This in turn improves responsiveness and tightens the feedback between the VSCode and the TypeScript Language Server. 
+
+## Migrating a codebase
+
+TypeScript's documentation and guidance on [Project References](https://www.typescriptlang.org/docs/handbook/project-references.html) are trivial to implement in new codebases. In larger codebases, leveraging Project References can be cumbersome. This section outlines our best practices on migrating a large codebase to gradually adopt Project References.
+
+### Understanding your codebase
+
+In most large codebases, migrating to Project References in one go is likely, not feasible. The migration would have to be done gradually over time. 
+
+We recommend starting out but identifying the leaf nodes in your project's depdency graph. These leaf nodes generally include the most widely used parts of your codebase. A good place to start might be the `packages/`, `tests/` or `app/utilities`. 
+
+In our experience, we've noticed that some `app/utilties` may be tightly coupled with `tests` utilties, `app` code and `packages/` . In some cases even resulting in circular dependencies. We suggest you identifying some of the these utilties and extracting them into isolated project references hosted in `packages/@<project>-utilties`.
+
+### Starting at the leaf nodes
+
+Start by creating an entrypoint for your project references. For simplicity, let's assume the leave nodes of your app consists of a `packages/` and `tests` directory. 
+
+```json
+// config/typescript/project-references.tsconfig.json
+
+{
+  "extends": "./tsconfig.base.json",
+  "files": [],
+  "include": [],
+  "references": [{"path": "../../packages"}, {"path": "../../tests"}]
+}
+
+// config/typescript/tsconfig.base.json
+
+{
+  "extends": "@shopify/typescript-configs/application.json",
+  "compilerOptions": {
+    "composite": true,
+    "noEmit": false,
+    // remaining default config
```

Comment:
[nitpick] the markdown editor/formatter marks this as invalid JSON :)

Reply:
good catch, needed `jsonc` instead.